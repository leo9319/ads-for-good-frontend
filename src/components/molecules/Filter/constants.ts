export const countryFilterOptions = [
  {
    name: 'India',
    code: 'india',
  },
  {
    name: 'USA',
    code: 'usa',
  },
  {
    name: 'Indonesia',
    code: 'indonsia',
  },
  {
    name: 'England',
    code: 'england',
  },
  {
    name: 'Africa',
    code: 'africa',
  },
];

export const genderFilterOptions = [
  {
    name: 'Boy',
    code: 'boy',
  },
  {
    name: 'Girl',
    code: 'girl',
  },
];

export const ageFilterOptions = [
  {
    name: '1-10',
    code: '1to10',
  },
  {
    name: '11-19',
    code: '11to19',
  },
  {
    name: '20-27',
    code: '20to27',
  },
  {
    name: '28-30',
    code: '28to30',
  },
];
