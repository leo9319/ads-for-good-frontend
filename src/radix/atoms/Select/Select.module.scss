@import '../../../components/molecules/UserProfile/UserProfile.scss';
@import '../../../styles/colours';

.selectWrapper {
  display: flex;
  flex-direction: column;
  max-width: 100%;
  width: 100%;
}

.requiredAsterisk {
  color: $color-red-600;
  position: absolute;
  margin-left: 25px;
  margin-bottom: 25px;
  width: var(--spacing-xs);
  height: var(--spacing-lg);
  font-size: var(--typography-font-size-5);

  &.size-1 {
    margin-left: 18px;
    margin-bottom: 5px;
  }
}

.customSelectTrigger {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: 100%;
  max-width: 100%;
  padding: 0 var(--spacing-md);
  border: 1px solid $color-neutral-300;
  background-color: $color-white;
  font-size: var(--typography-font-size-2);
  cursor: pointer;
  gap: var(--spacing-xs);
  height: 60px;
  border-radius: var(--spacing-sm);

  // Prevent the trigger from overflowing its container
  overflow: hidden;
  white-space: nowrap;

  &[data-placeholder] {
    > span:first-of-type {
      color: $color-neutral-600 !important;
    }
  }

  &.size-1 {
    height: 32px;
    border-radius: var(--spacing-xs);

    > span:first-of-type {
      font-size: var(--typography-font-size-2) !important;
      line-height: var(--typography-line-height-2) !important;
    }
  }

  // Remove default overflow logic from here; delegate to children
  > span:first-of-type {
    @include apply-text-styles(
      --typography-font-size-4,
      --typography-line-height-4,
      --typography-font-weight-regular,
      $color-neutral-950
    );

    flex: 1;
    min-width: 0 !important; // IMPORTANT: enables truncation with flex
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    text-align: start;
    display: inline-block;
    width: var(--radix-select-trigger-width);
  }

  &:not(:disabled) {
    &:hover {
      border: 1px solid $color-blue-400;
    }

    &:focus {
      border: 1px solid $color-blue-500;
      box-shadow: 0 0 0 4px $color-button-shadow;
    }

    &:active {
      border: 1px solid $color-blue-500;
    }

    &:focus-visible {
      outline: none;
    }

    &.error {
      border: 1px solid $color-red-600;
    }
  }

  &:disabled {
    background-color: $color-neutral-50;
    cursor: not-allowed;
  }
}

.customSelectIcon {
  display: flex;
  align-items: center;
  gap: var(--spacing-2xs);
  flex-shrink: 0;
  color: $color-neutral-950;

  .arrowDown {
    &.size-1 {
      span:first-of-type {
        width: 16px !important;
        height: 16px !important;
      }
    }

    span:first-of-type {
      width: 24px !important;
      height: 24px !important;
    }

    &.opened {
      rotate: 180deg;
      transition: rotate ease 0.3s;
    }
  }
}

.customSelectContent {
  width: var(--radix-select-trigger-width);
  max-width: 100%;
  background: $color-white;
  border-radius: var(--spacing-sm);
  padding: var(--spacing-xs);
  box-shadow: 0px 2px 4px -2px $color-box-shadow;
  position: relative;
  overflow: hidden;
  text-overflow: ellipsis;
  z-index: 100;
  margin: 0 !important;
  max-height: var(--radix-popper-available-height);
  height: 100%;
}

.customSelectViewport {
  gap: var(--spacing-2xs);
  display: grid;
  padding-right: 5px;
}

.customSelectItem {
  @include apply-text-styles(
    --typography-font-size-2,
    --typography-line-height-2,
    --typography-font-weight-regular,
    $color-neutral-950
  );

  padding: 0 var(--spacing-xs);
  cursor: pointer;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  height: 40px;
  border-radius: var(--spacing-xs);
  align-content: center;

  &:hover {
    background-color: $color-neutral-50;
  }

  &:focus-visible {
    outline: none;
  }

  &:not([data-state='checked']):focus-visible {
    background-color: $color-neutral-50;
  }

  &[data-state='checked'] {
    background-color: $color-neutral-950;
    color: $color-white !important;
    display: flex;
    align-items: center;
    gap: var(--spacing-2xs);
  }

  &[aria-disabled] {
    color: $color-neutral-600 !important;

    &:hover {
      background-color: $color-white;
      cursor: not-allowed;
    }
  }

  span:first-of-type {
    padding-left: 0;
  }

  &:not(:has(.itemIndicator)) span:first-of-type {
    padding-left: 20px;
  }

  &.size-1 {
    height: 32px;
    font-size: var(--spacing-sm);
    padding: 0 var(--spacing-2xs);
  }
}

.scrollContainer {
  > [data-radix-scroll-area-viewport] > div:first-of-type {
    display: flex !important;
  }
}
