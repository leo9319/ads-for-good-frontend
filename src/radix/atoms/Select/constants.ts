export const selectOptions = [
  {
    label: '2025',
    value: '2025',
    type: 'item',
    disabled: false,
  },
  {
    label: '2024',
    value: '2024',
    type: 'item',
    disabled: true,
    needSeparator: true,
  },
  {
    label: '2023',
    value: '2023',
    type: 'item',
    disabled: false,
  },
  {
    label: '2022',
    value: '2022',
    type: 'item',
    disabled: false,
  },
  {
    label: '2021',
    value: '2021',
    type: 'item',
    disabled: false,
  },
];
