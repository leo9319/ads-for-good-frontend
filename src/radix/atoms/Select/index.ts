export type { SelectProps } from './Select';
export { Select } from './Select';
export { default } from './Select';
