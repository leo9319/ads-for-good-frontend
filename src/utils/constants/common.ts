export const AJAX_CONST = {
  timout: 10000,
  algorithm: 'RSA-OAEP',
};
export const idRegexPattern = /(\{\(\[)([a-zA-Z0-9_]+)(\]\)\})/g;
export const idRegexPatternRemoval = /(\{\(\[|\]\)\})/g;
