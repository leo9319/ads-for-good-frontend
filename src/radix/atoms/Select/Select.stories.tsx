import React, { useState } from 'react';
import { Meta, StoryObj } from '@storybook/react';
import Theme from '@radix-styles/atoms/Theme';
import Select, { SelectProps } from './Select';
import { selectOptions } from './constants';

const WrapperComponent = ({
  name,
  size,
  required,
  disabled,
  options,
  placeholder,
}: SelectProps) => {
  const [value, setValue] = useState('');

  return (
    <Select
      name={name}
      size={size}
      placeholder={placeholder}
      required={required}
      disabled={disabled}
      options={options}
      value={value}
      onChange={setValue}
    />
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Library/Atoms/Select',
  component: WrapperComponent,
  decorators: [
    Story => (
      <Theme style={{ minWidth: '200px' }}>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/api/argtypes
  argTypes: {
    name: {
      control: 'text',
    },
    placeholder: {
      control: 'text',
    },
    size: {
      control: 'select',
      options: ['1', '2'],
    },
    required: {
      control: 'boolean',
    },
    disabled: {
      control: 'boolean',
    },
    options: {
      control: 'object',
    },
    error: {
      control: 'object',
    },
    status: {
      control: 'number',
    },
    onChange: {
      control: 'object',
    },
    value: {
      control: 'text',
    },
  },
  // more on args: https://storybook.js.org/docs/writing-stories/args#args-object
  args: {
    placeholder: 'Select option...',
  },
} satisfies Meta<typeof Select>;

export default meta;

type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/writing-stories/args

export const Default: Story = {
  args: {
    name: 'Service',
    size: '2',
    required: true,
    disabled: false,
    options: selectOptions,
    value: '',
    onChange: () => {},
  },
};
