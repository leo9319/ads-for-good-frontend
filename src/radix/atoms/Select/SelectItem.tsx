import React, { ReactElement } from 'react';
import { Select } from 'radix-ui';
import Icon from '@components/atoms/Icons';
import { classNames } from '@utils/common/classNames';
import styles from './Select.module.scss';

/**
 * @interface
 * Props for an individual selectable item inside a custom Select dropdown.
 */
interface SelectItemProps {
  /**
   * Text content displayed for the item.
   */
  children: string;

  /**
   * Value associated with this select item.
   */
  value: string;

  /**
   * Custom CSS class for styling the item.
   */
  className?: string;

  /**
   * Disables selection of the item if set to true.
   */
  disabled?: boolean;
}

export const SelectItem = React.forwardRef<HTMLDivElement, SelectItemProps>(
  (
    { children, className, value, disabled = false, ...props }: SelectItemProps,
    ref
  ): ReactElement => {
    return (
      <Select.Item
        value={value}
        disabled={disabled}
        className={classNames(styles.customSelectItem, className)}
        {...props}
        ref={ref}
      >
        <Select.ItemIndicator className={styles.itemIndicator}>
          <Icon name="icon-check" iconWidth="16" iconHeight="16" />
        </Select.ItemIndicator>
        <Select.ItemText>{children}</Select.ItemText>
      </Select.Item>
    );
  }
);

SelectItem.displayName = 'SelectItem';

export default SelectItem;
