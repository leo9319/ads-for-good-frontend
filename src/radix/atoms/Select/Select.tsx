import React, {
  Fragment,
  ReactElement,
  useEffect,
  useRef,
  useState,
} from 'react';
import { Select as RadixSelect } from 'radix-ui';
import { FieldError, FieldErrors } from 'react-hook-form';
import TextCaption from '@components/atoms/TextCaption';
import Icon from '@components/atoms/Icons';
import { classNames } from '@utils/common/classNames';
import Divider from '../Divider';
import ScrollArea from '../ScrollArea';
import SelectItem from './SelectItem';

import styles from './Select.module.scss';

/**
 * @interface
 * The structure of an individual option in a select dropdown
 */
interface SelectOptions {
  /**
   * Display label shown in the dropdown
   * @property
   */
  label: string;

  /**
   * Value associated with the select option
   * @property
   */
  value: string;

  /**
   * Type of the select option (e.g., "item", "group")
   * @property
   */
  type: string;

  /**
   * Whether the option is disabled and not selectable
   * @property
   */
  disabled?: boolean;

  /**
   * Whether a visual separator should follow this item
   * @property
   */
  needSeparator?: boolean;
}

/**
 * @interface
 * The props for the Select component used to render dropdowns
 */
export interface SelectProps {
  /**
   * Placeholder text displayed when no option is selected
   * @property
   */
  placeholder: string;

  /**
   * The name of the form field, used in form data
   * @property
   */
  name: string;

  /**
   * Array of options available for selection
   * @property
   */
  options: SelectOptions[];

  /**
   * Whether the select field is required
   * @property
   */
  required?: boolean;

  /**
   * The visual size variant of the select component (e.g., '1' or '2')
   * @property
   */
  size?: '1' | '2';

  /**
   * Callback triggered when a selection is made
   * @property
   */
  onChange: ([key]: string) => void;

  /**
   * The currently selected value
   * @property
   */
  value: string;

  /**
   * Error object containing field validation errors
   * @property
   */
  error?: { [key: string]: FieldError } | FieldErrors;

  /**
   * Status code for handling visual feedback or validation
   * @property
   */
  status?: number;

  /**
   * Whether the select input is disabled
   * @property
   */
  disabled?: boolean;

  /**
   * Additional class names for customizing styles
   * @property
   */
  containerClass?: string;
}

export const Select = ({
  value,
  status,
  options,
  onChange,
  size = '2',
  containerClass,
  error = {},
  name = 'select',
  required = false,
  disabled = false,
  placeholder = 'Select one...',
}: SelectProps): ReactElement => {
  const wrapperRef = useRef<HTMLDivElement>(null);
  const [isOptionsOpened, setIsOptionsOpened] = useState<boolean>(false);

  const mode = status && status >= 200 && status < 300 ? 'success' : 'error';

  const handleClick = (data: boolean) => {
    setIsOptionsOpened(data);
  };

  useEffect(() => {
    const wrapper = wrapperRef.current;
    if (!wrapper) return;

    const span = wrapper.querySelector(
      '.customSelectTrigger > span:first-of-type'
    ) as HTMLSpanElement | null;

    if (!span) return;

    const applyMaxWidth = () => {
      const maxTextWidth = wrapper.offsetWidth - 70;
      if (maxTextWidth > 0) {
        Object.assign(span.style, {
          maxWidth: `${maxTextWidth}px`,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap',
          display: 'inline-block',
        });
      }
    };

    const resizeObserver = new ResizeObserver(applyMaxWidth);
    resizeObserver.observe(wrapper);

    applyMaxWidth();

    return () => {
      resizeObserver.disconnect();
    };
  }, []);

  const concatSize = `size-${size}`;

  return (
    <div
      className={classNames(styles.selectWrapper, concatSize, containerClass)}
      id={`${name}-container`}
      ref={wrapperRef}
    >
      <RadixSelect.Root
        name={name}
        value={value}
        disabled={disabled}
        onOpenChange={handleClick}
        onValueChange={onChange}
      >
        <RadixSelect.Trigger
          className={classNames(
            styles.customSelectTrigger,
            concatSize,
            mode === 'error' && error[name] ? mode : ''
          )}
          aria-label={name}
          aria-labelledby={`${name}-label`}
        >
          <RadixSelect.Value placeholder={placeholder} />
          <RadixSelect.Icon className={styles.customSelectIcon}>
            <Icon
              name="icon-chevron-down"
              className={classNames(
                styles.arrowDown,
                concatSize,
                isOptionsOpened ? 'opened' : ''
              )}
            />
            {required && (
              <span className={classNames(styles.requiredAsterisk, concatSize)}>
                *
              </span>
            )}
          </RadixSelect.Icon>
        </RadixSelect.Trigger>
        <RadixSelect.Portal>
          <RadixSelect.Content
            className={styles.customSelectContent}
            position="popper"
            sideOffset={4}
          >
            <ScrollArea
              orientation="vertical"
              classname={styles.scrollContainer}
            >
              <RadixSelect.Viewport className={styles.customSelectViewport}>
                {options?.map(
                  ({ value, label, disabled, type, needSeparator }, idx) => {
                    if (type === 'item') {
                      return (
                        <Fragment key={`option-${idx + 1}`}>
                          <SelectItem
                            value={value}
                            disabled={disabled || false}
                            className={concatSize}
                          >
                            {label}
                          </SelectItem>
                          {needSeparator && (
                            <Divider
                              color="primary"
                              orientation="horizontal"
                              size={4}
                            />
                          )}
                        </Fragment>
                      );
                    }
                  }
                )}
              </RadixSelect.Viewport>
            </ScrollArea>
          </RadixSelect.Content>
        </RadixSelect.Portal>
      </RadixSelect.Root>
      {error[name] && !isOptionsOpened && (
        <TextCaption
          messages={(error[name] as FieldError)?.message}
          mode={mode}
        />
      )}
    </div>
  );
};

export default Select;
