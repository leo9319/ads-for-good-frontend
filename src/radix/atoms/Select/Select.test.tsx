import React from 'react';
import { render } from '@testing-library/react';
import Select from './Select';
import { selectOptions } from './constants';

describe('Select', () => {
  beforeAll(() => {
    Element.prototype.scrollIntoView = jest.fn();
    Element.prototype.scrollTo = jest.fn();

    Object.defineProperty(
      globalThis as unknown as { ResizeObserver?: jest.Mock },
      'ResizeObserver',
      {
        writable: true,
        configurable: true,
        value: jest.fn(() => ({
          observe: jest.fn(),
          unobserve: jest.fn(),
          disconnect: jest.fn(),
        })),
      }
    );
  });

  afterAll(() => {
    jest.restoreAllMocks();
    delete (globalThis as unknown as { ResizeObserver?: jest.Mock })
      .ResizeObserver;
  });

  it('Should Default Variant render successfully', () => {
    const { baseElement } = render(
      <Select
        name="Service"
        onChange={() => {}}
        options={selectOptions}
        placeholder="Select option..."
        required
        size="2"
        value="2023"
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
