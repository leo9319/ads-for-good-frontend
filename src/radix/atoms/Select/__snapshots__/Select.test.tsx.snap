// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Should Default Variant render successfully 1`] = `
<body>
  <div>
    <div
      class="select-wrapper size-2"
      id="Service"
    >
      <button
        aria-autocomplete="none"
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-label="Sample"
        class="custom-select-trigger size-2"
        data-state="closed"
        dir="ltr"
        role="combobox"
        type="button"
      >
        <span
          style="pointer-events: none;"
        >
          2023
        </span>
        <span
          aria-hidden="true"
          class="custom-select-icon"
        >
          <span
            class="arrow-down size-2"
            style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
          >
            <span
              class="icon icon-chevron-down"
            />
          </span>
          <span
            class="required-asterisk size-2"
          >
            *
          </span>
        </span>
      </button>
    </div>
  </div>
</body>
`;
